elastic/kibana
# [Synthetics] enable auto re-generation of monitor management api when read permissions are missing

Resolves https://github.com/elastic/kibana/issues/151695

Auto regenerates the synthetics api key when it does not include `synthetics-*` read permissions.

Also ensures key are regenerated when deleted via stack management. 

A user without permissions to enable monitor management will see this callout when monitor management is disabled for either reason
![Synthetics-Overview-Synthetics-Kibana (1)](https://user-images.githubusercontent.com/11356435/232926046-ea39115b-acc7-40a7-8ec1-de77a20daf53.png)

## Testing lack of read permissions

This PR is hard to test. I did so by adjusting the code to force the creation of an api key without read permissions. Here's how I did it:
1. connect to a clean ES instance by creating a new oblt cluster or running `yarn es snapshot
2. Remove read permissions for the api key https://github.com/elastic/kibana/pull/155203/files#diff-e38e55402aedfdb1a8a17bdd557364cd3649e1590b5e92fb44ed639f03ba880dR30
3. Remove read permission check here https://github.com/elastic/kibana/pull/155203/files#diff-e38e55402aedfdb1a8a17bdd557364cd3649e1590b5e92fb44ed639f03ba880dR60
4. Navigate to Synthetics app and create your first monitor
5. Navigate to Stack Management -> Api Keys. Click on he api key to inspect it's privileges. You should not see `read` permissions.
6. Remove the changes listed in step 2 and 3 and make sure the branch is back in sync with this PR
7. Navigate to the Synthetics app again.
9. Navigate to stack management -> api keys. Ensure there is only one synthetics monitor management api key. Click on he api key to inspect it's privileges. You should now see `read` permissions.
10. Delete this api key
11. Navigate back to the Synthetics app
12. Navigate back to stack management -> api keys. Notice tha api key has been regenerated


## Release note

If Monitor Management was enabled prior to 8.6.0, the API key generated internally will not contain the required permissions. The Synthetics app will attempt to fix this automatically when a user with sufficient privileges visits this page for the first time after upgrading to 8.8.0. Note however that any monitors configured to run on Elastic’s global managed testing infrastructure will stop running until a user with permissions has loaded the Synthetics app.
